### PR TITLE
FIX: code "block" detection before showing autocomplete

### PR DESCRIPTION
**Pull request by ZogStriP**
**TL;DR:** Refactor autocomplete to use async markdown parsing for code block detection.

Previously, the `inCodeBlock` function in `discourse/app/lib/utilities.js` used regular expressions to determine if a given position in the text was inside a code block. This approach had some limitations and could lead to incorrect behavior in certain edge cases.

This commit refactors `inCodeBlock` to use a more robust algorithm that leverages Discourse's markdown parsing library. 

The new approach works as follows:

1. Check if the text contains any code block markers using a regular expression.
   If not, return `false` since the cursor can't be in a code block.
1. If potential code blocks exist, find a unique marker character that doesn't appear in the text.
1. Insert the unique marker character into the text at the cursor position.
1. Parse the modified text using Discourse's markdown parser, which converts the markdown into a tree of tokens.
1. Traverse the token tree to find the token that contains the unique marker character.
1. Check if the token's type is one of the types representing code blocks ("code_inline", "code_block", or "fence").
   If so, return `true`, indicating that the cursor is inside a code block. 
   Otherwise, return `false`.

This algorithm provides a more accurate way to determine the cursor's position in relation to code blocks, accounting for the various ways code blocks can be represented in markdown.

To accommodate this change, the autocomplete `triggerRule` option is now an async function. 

The autocomplete logic in `composer-editor.js`, `d-editor.js`, and `hashtag-autocomplete.js` has been updated to handle the async nature of `inCodeBlock`.

Additionally, many of the tests have been refactored to handle async behavior. The test helpers now simulate typing and autocomplete selection in a more realistic, step-by-step manner. This should make the tests more robust and reflective of real-world usage.

This is a significant refactor that touches multiple parts of the codebase, but it should lead to more accurate and reliable autocomplete behavior, especially when dealing with code blocks in the editor.

> Written by an 🤖 LLM. Edited by a 🧑‍💻 human.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/d-editor.js [hunk 1]
```diff
@@ -532,10 +532,6 @@ export default Component.extend(TextareaTextManipulation, {
       },
 
       onKeyUp: (text, cp) => {
-        if (inCodeBlock(text, cp)) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/components/d-editor.js:535]: Not sure what this `onKeyUp` function is doing 🤔 I removed it and couldn't see any difference in behaviour locally.

File: app/assets/javascripts/discourse/app/lib/autocomplete.js [hunk 2]
```diff
@@ -537,19 +537,19 @@ export default function (options) {
     closeAutocomplete();
   });
 
-  function checkTriggerRule(opts) {
-    return options.triggerRule ? options.triggerRule(me[0], opts) : true;
+  async function checkTriggerRule(opts) {
+    return (await options.triggerRule?.(me[0], opts)) ?? true;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/autocomplete.js:541]: The most important change of this file is adding the `await` here.

File: app/assets/javascripts/discourse/app/lib/utilities.js [hunk 3]
```diff
@@ -421,34 +422,40 @@ export function postRNWebviewMessage(prop, value) {
...
-    end = match.index + match[0].length;
-    if (match.index <= pos && pos <= end) {
-      return true;
-    }
+const CODE_TOKEN_TYPES = ["code_inline", "code_block", "fence"];
+
+export async function inCodeBlock(text, pos) {
+  const marker = pickMarker(text);
+  if (!marker) {
+    return false;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/utilities.js:451]: If we can't find a unique character to use as a marker, then we bail out. We probably have bigger problems than showing the autocomplete 😅

File: app/assets/javascripts/discourse/tests/unit/lib/utilities-test.js [hunk 4]
```diff
@@ -219,34 +219,16 @@ module("Unit | Utilities", function (hooks) {
...
-    ];
-
-    texts.forEach((text) => {
-      for (let i = 0; i < text.length; ++i) {
-        if (text[i] === "0" || text[i] === "1") {
-          assert.strictEqual(inCodeBlock(text, i), text[i] === "1");
-        }
+  test("inCodeBlock", async function (assert) {
+    const text =
+      "000\n\n```\n111\n```\n\n000\n\n`111 111`\n\n000\n\n[code]\n111\n[/code]\n\n    111\n\t111\n\n000`000";
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/unit/lib/utilities-test.js:224]: I think that one test is plenty enough to cover most of the cases. Since it's re-using the same Markdown engine as the composer, we can safely rely on it being correct when identifying code blocks.

**Comment by ZogStriP** [comment]
> * I probably broke some autocomplete tests with my changes, but I can't seem to be able to run the `qunit` tests locally - all my browsers (Safari, Chrome, and Firefox) crash after ~2000 tests 💥.

Welp, I did 😅 Will have a look tomorrow.

File: app/assets/javascripts/discourse/app/lib/autocomplete.js [hunk 5]
```diff
@@ -537,19 +540,20 @@ export default function (options) {
     closeAutocomplete();
   });
 
-  function checkTriggerRule(opts) {
-    return options.triggerRule ? options.triggerRule(me[0], opts) : true;
+  async function checkTriggerRule(opts) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/autocomplete.js:543]: Making `checkTriggerRule` async forced all the other `async`/`await` changes in this file.

File: app/assets/javascripts/discourse/app/lib/utilities.js [hunk 6]
```diff
@@ -421,34 +422,45 @@ export function postRNWebviewMessage(prop, value) {
...
+    if (!text.includes(char)) {
+      return char;
+    }
+  }
+  return null;
+}
 
-const OPEN_CODE_BLOCKS_REGEX = /^(  |\t).*|`[^`]+|^```[^]*?|\[code\][^]*?/gm;
+function findToken(tokens, marker, level = 0) {
+  if (level > 50) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/utilities.js:439]: Stack overflow protection. Deeper than 50 level? Just bail out.

File: app/assets/javascripts/discourse/app/lib/utilities.js [hunk 7]
```diff
@@ -421,34 +422,45 @@ export function postRNWebviewMessage(prop, value) {
...
-  for (const match of text.matchAll(CODE_BLOCKS_REGEX)) {
-    end = match.index + match[0].length;
-    if (match.index <= pos && pos <= end) {
-      return true;
-    }
+const CODE_MARKERS_REGEX = /    |```|~~~|(?<!`)`(?!`)|\[code\]/;
+const CODE_TOKEN_TYPES = ["code_inline", "code_block", "fence"];
+
+export async function inCodeBlock(text, pos) {
+  if (!CODE_MARKERS_REGEX.test(text)) {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/utilities.js:450]: Quickly look for code markers so we can bail out if there aren't any.

File: app/assets/javascripts/discourse/tests/acceptance/composer-editor-mentions-test.js [hunk 8]
```diff
@@ -65,11 +61,14 @@ acceptance("Composer - editor mentions", function (needs) {
     await visit("/");
     await click("#create-topic");
 
-    await emulateAutocomplete(".d-editor-input", "abc @u");
-    await click(".autocomplete.ac-user .selected");
+    const editor = query(".d-editor-input");
+
+    for (let c of "abc @u\r") {
+      await simulateKey(editor, c);
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/acceptance/composer-editor-mentions-test.js:67]: Replaced the hacky `emulateAutocomplete` with `simulateKey`.

Also replaced the `await click(".autocomplete.ac-user .selected");` with emulating pressing <kbd>Enter</kbd> (aka "\r") so it's (hopefully) more reliable and faster.

File: app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js [hunk 9]
```diff
@@ -610,14 +605,28 @@ export async function paste(element, text, otherClipboardData = {}) {
...
-  await fillIn(inputSelector, text);
-  await triggerKeyEvent(inputSelector, "keyup", "Backspace");
+  const pos = element.selectionStart;
+  const value = element.value;
+
+  if (key === "\b") {
+    element.value = value.slice(0, pos - 1) + value.slice(pos);
+    element.selectionStart = pos - 1;
+    element.selectionEnd = pos - 1;
+  } else if (key !== "\r") {
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js:621]: To better simulate the behaviour that is happening in the browser, we don't "add" the <kbd>Enter</kbd> character in the `element`.

It will be caught by the `autocomplete` library though.

File: app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js [hunk 10]
```diff
@@ -610,14 +605,28 @@ export async function paste(element, text, otherClipboardData = {}) {
...
+    element.value = value.slice(0, pos) + key + value.slice(pos);
+    element.selectionStart = pos + 1;
+    element.selectionEnd = pos + 1;
+  }
+
+  element.dispatchEvent(new KeyboardEvent("keyup", event));
+
+  await settled();
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js:629]: That was the missing piece 🙏

File: app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js [hunk 11]
```diff
@@ -1,154 +1,85 @@
...
-import autocomplete from "discourse/lib/autocomplete";
+import { simulateKey } from "discourse/tests/helpers/qunit-helpers";
 
 module("Unit | Utility | autocomplete", function (hooks) {
   setupTest(hooks);
   let elements = [];
 
-  function textArea(value) {
-    let element = document.createElement("TEXTAREA");
+  const template = compile(
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js:12]: We're always using the same `template` so I extracted it so we only compile it once.

File: app/assets/javascripts/discourse/app/lib/autocomplete.js [hunk 12]
```diff
@@ -17,8 +17,9 @@ import discourseLater from "discourse-common/lib/later";
 
 export const SKIP = "skip";
 export const CANCELLED_STATUS = "__CANCELLED";
-const allowedLettersRegex = /[\s\t\[\{\(\/]/;
-let _autoCompletePopper;
+
+const ALLOWED_LETTERS_REGEXP = /[\s[{(/]/;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/autocomplete.js:21]: The `\t` is already included in `\s`.
Also removed the `/` used to escape the characters which isn't required in a `[]`.

File: app/assets/javascripts/discourse/app/lib/autocomplete.js [hunk 13]
```diff
@@ -17,8 +17,9 @@ import discourseLater from "discourse-common/lib/later";
...
+let _autoCompletePopper, _inputTimeout;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/autocomplete.js:22]: Added a `_` prefix to `inputTimeout` so it better indicates it's only used internally.

File: app/assets/javascripts/discourse/app/lib/autocomplete.js [hunk 14]
```diff
@@ -374,21 +373,26 @@ export default function (options) {
     } else {
       selectedOption = -1;
     }
-    ul.find("li").click(function ({ originalEvent }) {
+    ul.find("li").click(async function ({ originalEvent }) {
+      // this is required to prevent the default behaviour when clicking on a <a> tag
+      originalEvent.preventDefault();
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/app/lib/autocomplete.js:378]: Had to `preventDefault()` otherwise the page would reload if you were to click on a autocomplete item 😬

File: app/assets/javascripts/discourse/tests/helpers/component-test.js [hunk 15]
```diff
@@ -12,6 +12,7 @@ import { currentSettings } from "discourse/tests/helpers/site-settings";
 
 export function setupRenderingTest(hooks) {
   emberSetupRenderingTest(hooks);
+  let _oldAutocomplete;
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/helpers/component-test.js:15]: Oh my f*cking gawd! That was the part that was _randomly_ breaking the `rake qunit:test` depending on whether a test calling this `setupRenderingTest` method would run **before** any autocomplete tests, since this was basically nullifying the `autocomplete` "plugin".

Saving the method and restoring it once the "component test" is done made it all work 🎉 

(cc @CvX - that was from a [commit](https://github.com/discourse/discourse/commit/f1136481078b6f3e12f6c571134830847e96250d) from 3 years ago 😅)

File: app/assets/javascripts/discourse/app/lib/autocomplete.js (hunk 14 above)
Comment (CvX) [reply app/assets/javascripts/discourse/app/lib/autocomplete.js:378]: Yeah, previously the function was returning `false` -  which is jQuery for "prevent the default" but now we're returning a promise so this no longer works.

which also means you can remove the `return false` at the bottom (line 400) 🙂

File: app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js [hunk 16]
```diff
@@ -157,105 +84,81 @@ module("Unit | Utility | autocomplete", function (hooks) {
...
+    await simulateKey(element, "@");
 
-    let list = document.querySelectorAll("#ac-testing ul li");
-    assert.strictEqual(list.length, 2);
+    assert.strictEqual(queryAll("#ac-testing ul li").length, 2);
 
-    let selected = document.querySelectorAll("#ac-testing ul li a.selected");
+    const selected = queryAll("#ac-testing ul li a.selected");
     assert.strictEqual(selected.length, 1);
     assert.strictEqual(selected[0].innerText, "test1");
```
Comment (CvX) [inline app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js:163]:
```suggestion
    assert.dom("#ac-testing li a.selected").exists({ count: 1 });
    assert.dom("#ac-testing li a.selected").hasText("test1");
```

File: app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js [hunk 17]
```diff
@@ -157,105 +84,81 @@ module("Unit | Utility | autocomplete", function (hooks) {
...
-</div>`),
     });
 
-    element.dispatchEvent(new KeyboardEvent("keydown", { key: "@" }));
-    element.dispatchEvent(new KeyboardEvent("keyup", { key: "@" }));
+    await simulateKey(element, "@");
 
-    let list = document.querySelectorAll("#ac-testing ul li");
-    assert.strictEqual(list.length, 2);
+    assert.strictEqual(queryAll("#ac-testing ul li").length, 2);
```
Comment (CvX) [inline app/assets/javascripts/discourse/tests/unit/lib/autocomplete-test.js:159]: qunit-dom

```suggestion
    assert.dom("#ac-testing ul li").exists({ count: 2 });
```

the same can be used in other places

File: app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js (hunk 9 above)
Comment (CvX) [reply app/assets/javascripts/discourse/tests/helpers/qunit-helpers.js:621]: This helper function can be simplified into:

```js
  if (key === "\b") {
    await triggerKeyEvent(element, "keydown", "Backspace");

    const pos = element.selectionStart;
    element.value = element.value.slice(0, pos - 1) + element.value.slice(pos);
    element.selectionStart = pos - 1;
    element.selectionEnd = pos - 1;

    await triggerKeyEvent(element, "keyup", "Backspace");
  } else if (key !== "\r") {
    await typeIn(element, key);
  } else {
    await triggerKeyEvent(element, "keydown", "Enter");
    await triggerKeyEvent(element, "keyup", "Enter");
  }
```

File: app/assets/javascripts/discourse/app/lib/utilities.js [hunk 18]
```diff
@@ -421,34 +422,45 @@ export function postRNWebviewMessage(prop, value) {
...
+}
 
-export function inCodeBlock(text, pos) {
-  let end = 0;
-  for (const match of text.matchAll(CODE_BLOCKS_REGEX)) {
-    end = match.index + match[0].length;
-    if (match.index <= pos && pos <= end) {
-      return true;
-    }
+const CODE_MARKERS_REGEX = /    |```|~~~|(?<!`)`(?!`)|\[code\]/;
```
[the remark this answers is not in the document]
Comment (ZogStriP) [reply app/assets/javascripts/discourse/app/lib/utilities.js:446]: Yeah, that was totally wrong (or rather a workaround due to the wrong behaviour added by the too simple regex).

A code block is **always** indented by at least 4 code blocks.

File: app/assets/javascripts/discourse/tests/helpers/component-test.js (hunk 15 above)
Comment (ZogStriP) [reply app/assets/javascripts/discourse/tests/helpers/component-test.js:15]: Nothing... So it's gone now 🔥

File: app/assets/javascripts/discourse/tests/helpers/component-test.js [hunk 19]
```diff
@@ -48,8 +47,6 @@ export function setupRenderingTest(hooks) {
 
     autoLoadModules(this.owner, this.registry);
     this.owner.lookup("service:store");
-
-    $.fn.autocomplete = function () {};
```
Comment (ZogStriP) [inline app/assets/javascripts/discourse/tests/helpers/component-test.js:52]: 🔥

**Review by CvX — APPROVED** [review]
